File: Tic_Tac_Toe.cpp
// Rosie Khurmi
// July 2023
// Tic_Tac_Toe will simulate a game of tic-tac-toe between 2 players 

#include "Tic_Tac_Toe.h"

// Parameterized Constructor, takes in size of board
TTT::TTT (int s) { 
    this->size = s;
    this->board.resize(this->size, std::vector<char>(this->size, '-'));
}

// Getters
std::vector<std::vector<char>> TTT::getBoard() {
    return this->board;
}

int TTT::getSize() {
    return this->size;
}

// Setters
void TTT::setBoard (int size) {
    this->size = size;
    this->board.resize(size, std::vector<char>(size, '-'));
}

// Equality Checkers

// checkDiagonalLeft() takes in the number of the player 
// Returns true if the top left to bottom right is the same, otherwise false
bool TTT::checkDiagonalLeft(int player) {

    int count = 0; // Counter for how many in a diagonal
    char compare; // Comaprison variable
    
    // Change comapre depending if its the turn of player 1 or player 2
    if (player == 1) {
        compare = 'X';
    } else if (player == 2) {
        compare = 'O';
    }

    // Go through rows
    for (int i = 0; i < this->size; i++) {
        // Go through columns
        for (int j = 0; j < this->size; j++) {
            // Increase counter if piece on left diagonals on board match compare
            if ((this->board[i][j] == compare) && (i == j)) {
                count ++;
            }

        }
    }

    // Return true if counter is the same size as board, otherwise false
    if (count == this->size) {
        return true;
    }
    return false;

}

// checkDiagonalLeft() takes in the number of the player 
// Returns true if the top right to bottom left is the same, otherwise false
bool TTT::checkDiagonalRight(int player) {

    int count = 0; // Counter for how many in a diagonal
    char compare; // Comaprison variable
    
    // Change comapre depending if its the turn of player 1 or player 2
    if (player == 1) {
        compare = 'X';
    } else if (player == 2) {
        compare = 'O';
    }

    // Go through rows
    for (int i = 0; i < this->size; i++) {
        // Go through columns
        for (int j = 0; j < this->size; j++) {
            // Increase counter if piece on right diagonals on board match compare
            if ((this->board[i][j] == compare) && (j == (this->size- 1) - i)) {
                count ++;
            }

        }
    }

    // Return true if counter is the same size as board, otherwise false
    if (count == this->size) {
        return true;
    }
    return false;

}

// checkRows() takes in the number of the player 
// Returns if all the same are in any row
bool TTT::checkRows(int player) { 

    int count = 0; // Counter for how many in a row
    char compare; // Comaprison variable
    
    // Change comapre depending if its the turn of player 1 or player 2
    if (player == 1) {
        compare = 'X';
    } else if (player == 2) {
        compare = 'O';
    }
    
    // Go through each row
    for (int i = 0; i < this->size; i++) {
        // Use this loop to check if each column of the row matches compare
        for (int j = 0; j < this->size; j++) {
            if (this->board[i][j] == compare) {
                count ++;
            }
        }
        // If counter equals total size in the row, return true
        if (count == this->size) {
            return true;
        // Else, set the count to 0 and repeat the process for the other rows
        } else {
            count = 0;
        }
    }

    // If no wins are found, return false
    return false;

}

// checkColumns() takes in the number of the player 
// Returns if all the same are in any row
bool TTT::checkColumns(int player) {

    int count = 0; // Counter for how many in a column
    char compare; // Comaprison variable
    
    // Comaprison variable
    if (player == 1) {
        compare = 'X';
    } else if (player == 2) {
        compare = 'O';
    }
    
    // Go through each column
    for (int i = 0; i < this->size; i++) {
        // Use this loop to check if each row of the column matches compare
        for (int j = 0; j < this->size; j++) {
            if (compare == this->board[j][i]) {
                count ++;
            }
        }

        // If counter equals total size in the row, return true
        if (count == this->size) {
            return true;
        // Else, set the count to 0 and repeat the process for the other rows
        } else {
            count = 0;
        }

    }

    // If no wins are found, return false
    return false;

}

// enterValue() takes in player number
// Adds the player piece to the game board
void TTT::enterValue(int player) {

    std::cout << "\nPlayer " << player << std::endl << std::endl; 

    int row, column;

    std::cout << "Enter a Column: ";
    std::cin >> column;

    std::cout << "Enter a Row: ";
    std::cin >> row;

    // If the user chosen square is empty, add the corresponding piece
    if (this->board[row - 1][column - 1] == '-') {
        if (player == 1) {
        this->board[row - 1][column - 1] = 'X';
        } if (player == 2) {
        this->board[row - 1][column - 1] = 'O';
        }
    }
    // Else, ask user to enter again
    else {
        std::cout << "\nInvalid Choice\n" << std::endl;
        enterValue(player);
    }

}

// checkFull() checks if the game board is full
// Returns true or false accordingly
bool TTT::checkFull() {

    for (int i = 0; i < this->size; i++) {
       for (int j = 0; j < this->size; j++) {
            if (this->board[i][j] == '-') {
                return false;
            }
        } 
    }

    return true;

}

// displayBoard() prints what is in the board at any given time
void TTT::displayBoard() {

    std::cout << "\nBOARD\n" << std::endl;

    for (int i = 0; i < this->size; i++) {
        for (int j = 0; j < this->size; j++) {
            std::cout << "   " << this->board[i][j] << "   ";
        } 
        std::cout << std::endl;
    }

}

// getSizeBoard() takes in a pointer to the game board size
/* Asks user what size board they want and modifies size to that value
if => 3 */
void getSizeBoard(int* size) {

    std::cout << "Enter the size of board you would like: ";
    std::cin >> *size;
    
    if (*size < 3) {
        std::cout << "The size of the board must be at least 3 units\n" << std::endl;
        getSizeBoard(size);
    }

}

// playGame() is a static method that takes in a TTT object and a pointer to the player
// Executes the tic-tac-toe game
void playGame(TTT game, int* player) {

        // Execute while the board is not full
        while (game.checkFull() != true) {
 
        // Using pointers allows player variable to be modifed in memory
        game.enterValue(*player);
        game.displayBoard();

        // If there is a match anyhwere, exit the method 
        if (game.checkDiagonalLeft(*player) || game.checkDiagonalRight(*player) 
        || game.checkColumns(*player) || game.checkRows(*player)) {
            std::cout << "\nPLAYER " << *player << " HAS WON THE GAME" << std::endl;
            return;
        // Else, switch players and execute loop again
        } else {
            if (*player == 1) {
                *player = 2;
            } else if (*player == 2) {
                *player = 1;
            }
        } 


    }

}

File: tictactoe.py
# Rosie Khurmi
# September 2, 2023

"""
Tic Tac Toe Player
"""

import math
import copy

X = "X"
O = "O"
EMPTY = None

# initial_state() returns an empty 3 x 3 board
def initial_state():

    return [
        [EMPTY, EMPTY, EMPTY],
        [EMPTY, EMPTY, EMPTY],
        [EMPTY, EMPTY, EMPTY]
    ]

# player() returns which player has the next turn on the board
# @param board: 3 x 3 board
def player(board):

    # Count the number of X's and O's on the board
    x_count = sum(row.count('X') for row in board)
    o_count = sum(row.count('O') for row in board)

    # If X = X, return X, else O
    player = X if x_count <= o_count else O
    return player

# actions() returns a set of all possible actions (i, j) available on the board
# @param board: 3 x 3 board   
def actions(board):

    options = set() # Hold all options in a set

    # Go through 2D array
    # If there is an empty spot, add its coordinates to the set
    for r in range(len(board)):
        for c in range(len(board)):
            if board[r][c] == EMPTY:
                options.add((r, c))

    return options # Return all possible option

# result() returns the board that results from making move (i, j) on the board
# @param board: 3 x 3 board, action: set of (i, j)
def result(board, action):

    # Create a deep copy of the board
    result = copy.deepcopy(board)

    r = action[0] # Row
    c = action[1] # Colum

    # If the sopt is empty, append the corresponding player
    # Else raise an exception
    if result[r][c] == EMPTY:
        result[r][c] = player(result)
        return result
    else:
        raise ValueError

# winner() returns the winner of the game, if there is one
# @param board: 3 x 3 board
def winner(board):

    winner = None

    # Check if top to bottom diagonal is all the same
    if board[0][0] == board[1][1] == board[2][2]:
        # Return X or O depending who gets all the same
        winner = X if board[0][0] == X else O if board[0][0] == O else None
        
    # Check if bottom to top diagonal is all the same
    if board[0][2] == board[1][1] == board[2][0]:
        # Return X or O depending who gets all the same
        winner = X if board[0][2] == X else O if board[0][2] == O else None
        
    # Check matching in any 3 rows
    for r in board: 
        if r[0] == r[1] == r[2]:
            # Return X or O depending who gets all the same
            winner = X if r[0] == X else O if r[0] == O else None

    # Check matching in any 3 columns     
    for c in range(len(board)): 
        if board[0][c] == board[1][c] == board[2][c]:
            # Return X or O depending who gets all the same
            winner = X if board[0][c] == X else O if board[0][c] == O else None
            
    return winner

# terminal() returns True if the game is over, False otherwise
# @param board: 3 x 3 board
def terminal(board):

    # If a winner has been declared or there are no spaces on the board
    # Return True, else False
    if (winner(board) or (len(actions(board)) == 0)):
        return True
    else:
        return False

# utility() returns 1 if X has won the game, -1 if O has won, 0 otherwise
# @param board: 3 x 3 board
def utility(board):

    # If board has reached terminal
    if terminal(board):
        # Return 1 if X has won
        if winner(board) == X:
            return 1 
        # Return -1 if O won
        elif winner(board) == O:
            return -1
        # Return 0 if tie
        else:
            return 0 

# optimal() returns the optimal action for the current player on the board
# @param board: 3 x 3 board, depth: int
def optimal(board, depth = 0):

    # If the board is at terminal, return the winner and no possible solution  
    if terminal(board):
        return utility(board), None
    
    # Else perform minimax
    else:

        move = set() # Hold best move

        if player(board) == X: # Maximizing Player
            best = float('-inf') # Optimal Value - Try to get to 1
    
            # Go through all empty spots
            for a in actions(board):
                # Perform a recursive call of the fuction with all possible spots
                temp, m = optimal(result(board, a))

                # If the result is a larger value, change best and store the move
                if temp > best:
                    best = temp
                    move = a
    
            # Return best and move
            return best, move
        
        elif player(board) == O: # Minimizing Player

            best = float('inf')  # Optimal Value - Try to get to -1

            # Go through all empty spots
            for a in actions(board):
                # Perform a recursive call of the fuction with all possible spots
                temp, m = optimal(result(board, a))

                # If the result is a smaller value, change best and store the move
                if temp < best:
                    best = temp
                    move = a

            # Return best and move
            return best, move

# minimax() returns the optimal action for the current player on the board
# @param board: 3 x 3 board
def minimax(board):

    # Call optimal and return the move
    value, move = optimal(board)
    return move

 









    



    











    









   


